sword2/utils.py

```diff
@@ -188,42 +185,47 @@ def create_multipart_related(payloads):
-    BOUNDARY = '===========%s_$' % bhash
-    CRLF = '\r\n'   # As some servers might barf without this.
+    bhash = md5(datetime.now().isoformat().encode('utf-8')).hexdigest()    # eg 'd8bb3ea6f4e0a4b4682be0cfb4e0a24e'
+    BOUNDARY = b'===========%s_$' % bhash.encode('utf-8')
+    CRLF = b'\r\n'   # As some servers might barf without this.
     body = []
     for payload in payloads:   # predicatable ordering...
-        body.append('--' + BOUNDARY)
+        body.append(b'--' + BOUNDARY)
         if payload.get('type', None):
-            body.append('Content-Type: %(type)s' % payload)
+            toappend = 'Content-Type: %(type)s' % payload
+            body.append(toappend.encode('utf-8'))
         else:
-            body.append('Content-Type: %s' % get_content_type(payload.get("filename")))
+            toappend = 'Content-Type: %s' % get_content_type(payload.get("filename"))
+            body.append(toappend.encode('utf-8'))
             
         if payload.get('filename', None):
-            body.append('Content-Disposition: attachment; name="%(key)s"; filename="%(filename)s"' % (payload))
+            toappend = 'Content-Disposition: attachment; name="%(key)s"; filename="%(filename)s"' % (payload)
+            body.append(toappend.encode('utf-8'))
         else:
-            body.append('Content-Disposition: attachment; name="%(key)s"' % (payload))
+            toappend = 'Content-Disposition: attachment; name="%(key)s"' % (payload)
+            body.append(toappend.encode('utf-8'))
         
-        if payload.has_key("headers"):
-            for f,v in payload['headers'].iteritems():
-                body.append("%s: %s" % (f, v))     # TODO force ASCII?
+        if "headers" in payload:
+            for f,v in payload['headers'].items():
+                toappend = "%s: %s" % (f, v)
+                body.append(toappend.encode('utf-8'))     # TODO force ASCII?
         
-        body.append('MIME-Version: 1.0')
+        body.append(b'MIME-Version: 1.0')
         if payload['key'] == 'payload':
-            body.append('Content-Transfer-Encoding: base64')
-            body.append('')
+            body.append(b'Content-Transfer-Encoding: base64')
+            body.append(b'')
             if hasattr(payload['data'], 'read'):
                 body.append(b64encode(payload['data'].read()))
             else:
-                body.append(b64encode(payload['data']))
+                body.append(b64encode(payload['data'].encode('utf-8')))
         else:
-            body.append('')
+            body.append(b'')
             if hasattr(payload['data'], 'read'):
                 body.append(b64encode(payload['data'].read()))
             else:
-                body.append(b64encode(payload['data']))
-    body.append('--' + BOUNDARY + '--')
-    body.append('')
+                body.append(b64encode(payload['data'].encode('utf-8')))
+    body.append(b'--' + BOUNDARY + b'--')
+    body.append(b'')
     body_bytes = CRLF.join(body)
```

Comment:
Instead of inserting bytes and encoding strings to bytes everywhere in here, you could just encode it here as you have one big string. It will be a little more readable without this `b'` strings and `.encode` methods everythere;)